### PR TITLE
WIP: Correct Identifier Tracking

### PR DESCRIPTION
**Pull request by mitsuhiko**
This branch implements correct identifier tracking through aliasing. This basically solves Jinja's oldest bug which is that it tries to trivially map identifiers onto Python scoping which cannot work because Python scoping is not expressive enough to support all of Jinja's features.

This will also fix a ton of other issues and open up heavily improved includes.

The main breakage this causes which is known is that code that overrides Context now needs to override `resolve_or_missing` instead of `resolve` if it wants custom resolve support.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
Open TODOs:

* [ ] we use the same scope for the normal loop bodies and else blocks. This generates non optimal resolves
* [ ] we generate some resolves that are strictly unnecessary. Maybe we can do some dead code detection here
* [ ] The logic for undefines changed so that undefined checks are now generated constantly on value use instead of on resolve. This changes `Context.resolve` to be an API that is no longer used by Jinja internally and might be less efficient. Requires some testing.

**Comment by mitsuhiko** [comment]
Bonus feature: no longer uses `locals()` which should greatly help pypy.

File: jinja2/compiler.py [hunk 1]
```diff
@@ -1073,15 +841,9 @@ def visit_FromImport(self, node, frame):
...
-        # when calculating the nodes for the inner frame we have to exclude
-        # the iterator contents from it
-        children = node.iter_child_nodes(exclude=('iter',))
-        if node.recursive:
-            loop_frame = self.function_scoping(node, frame, children,
-                                               find_special=False)
-        else:
-            loop_frame = frame.inner()
-            loop_frame.inspect(children)
+        # TODO: this shoudl really use two frames: one for the loop body
```
Comment (davidbgk) [inline jinja2/compiler.py]: Typo: should

**Comment by mitsuhiko** [comment]
This should be able to fix #110 now. Still failing but easy to fix.

**Pull request merged**